stdlib/public/core/Stride.swift

```diff
@@ -171,6 +224,39 @@ extension Strideable {
   }
 }
 
+extension Strideable where Self: FixedWidthInteger & SignedInteger {
+  @_alwaysEmitIntoClient
+  public static func _step(
+    after current: (index: Int?, value: Self),
+    from start: Self, by distance: Self.Stride
+  ) -> (index: Int?, value: Self) {
+    let value = current.value
+    let (partialValue, overflow) =
+      Self.bitWidth >= Self.Stride.bitWidth ||
+        (value < (0 as Self)) == (distance < (0 as Self.Stride))
+          ? value.addingReportingOverflow(Self(distance))
+          : (Self(Self.Stride(value) + distance), false)
```

Comment:
Yes, this is to mitigate the edge case where a valid `distance` between two signed integer values exceeds the representable range of `Self`. No such `distance` exists between two signed integer values of the same sign, or between two unsigned integer values, but they do exist between two signed integer values of opposite sign.

(Eliminating all possible out-of-bounds trapping--i.e., testing for an _invalid_ `distance`--would be wasteful to repeat for every step but cannot be done upfront in the generic context without causing the behavioral regression you mention since it would involve a potentially trapping `distance(to:)`. Fortunately, that's something that the user can test before they stride, and we can improve the documentation on that, since it's never worked without trapping.)

Reply:
Ah, I see! `Int8.Stride == Int` 🤯 

Note that `BinaryInteger.advanced(by:)` is implemented as `self + Self(distance)` -- if we consider the overflowing `distance` case important, then we probably ought to fix it there first. (I don't think it's overly important, though.)

In fact, why not just have this call `value.advanced(by: distance)` in the fallback case? That matches the default implementation we're replacing, and gives Strideable & FixedWidthInteger types an obvious way to fix the oversized distance issue.

Reply:
Not quite, `BinaryInteger.advanced(by:)` does exactly what we do here, and the reason not to all it is because we will have arrived at the fallback case only by already checking what `advanced(by:)` would check if we were to call it:

```
  @inlinable
  @inline(__always)
  public func advanced(by n: Int) -> Self {
    if !Self.isSigned {
      return n < (0 as Int)
        ? self - Self(-n)
        : self + Self(n)
    }
    if (self < (0 as Self)) == (n < (0 as Self)) {
      return self + Self(n)
    }
    return self.magnitude < n.magnitude
      ? Self(Int(self) + n)
      : self + Self(n)
  }
}
```

Reply:
Hah, so it does! Fascinating. OK, I'm fully convinced this is the right thing to do. Thank you for explaining all this! 😊